### PR TITLE
feat: Make custom metrics available as methods

### PR DESCRIPTION
**Pull request by auguste-probabl**
Closes https://github.com/probabl-ai/skore-hub/issues/941

The main changes are in the first commit, the rest is some cleanups I did along the way; I can split them into other PRs if need be.

Overrides `__getattr__` so that the metrics registry is checked when `report.metrics.<thing>` is accessed (except if the method is defined statically). I replaced `__getattribute__` with `__getattr__` because `__getattribute__` is more low-level and harder to manipulate safely. This made it harder to both add new methods and hide non-applicable ones, so I put back `available_if`s and made `__getattr__`'s only responsibility to add new methods. This works out nicely because every `_MetricsAccessor` class now shares the same behaviour and I could refactor it to a `BaseMetricsAccessor`.

As discussed IRL, re-defining the builtin metrics using this mechanism is deferred because the UX would be really bare-bones (no documentation).

### CODE REVIEW

File: skore/tests/unit/reports/cross_validation/metrics/test_numeric.py [hunk 1]
```diff
@@ -214,8 +215,10 @@ def test_precision_recall_pos_label_overwrite(
 
 def test_get(forest_binary_classification_data):
     """``get`` works."""
-    estimator, X, y = forest_binary_classification_data
-    report = CrossValidationReport(estimator, X, y, splitter=2)
+    _, X, y = forest_binary_classification_data
```
Comment (GaetandeCast) [inline skore/tests/unit/reports/cross_validation/metrics/test_numeric.py]:
```suggestion
    X, y = binary_classification_data
```

Comment (GaetandeCast) [reply skore/tests/unit/reports/cross_validation/metrics/test_numeric.py]: and everywhere else

File: skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py [hunk 2]
```diff
@@ -351,3 +351,26 @@ def test_get_custom(comparison_estimator_reports_binary_classification):
         "DummyClassifier_1": {"Hello": 1},
         "DummyClassifier_2": {"Hello": 1},
     }
+
+
+# report.metrics.<custom>
+
+
```
Comment (GaetandeCast) [inline skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py]:
```suggestion
```

Same in all 3 files

**Review by GaetandeCast — CHANGES_REQUESTED** [review]
Two nits on checks.

One pb I had that must be resolved is that I can't access a metric starting by `neg` but I can get it with `.get()`. Can be seen in the following snippet:

```python
from skore import evaluate
from sklearn.datasets import make_regression
from sklearn.ensemble import HistGradientBoostingRegressor

X, y = make_regression(n_samples=500, n_features=50, n_targets=1, random_state=0)
report = evaluate(HistGradientBoostingRegressor(random_state=0), X, y, splitter=5)
report.metrics.add("neg_mean_absolute_percentage_error")
report.metrics.get("neg_mean_absolute_percentage_error") # Works
report.metrics.neg_mean_absolute_percentage_error() # Fails
```

**Comment by thomass-dev** [comment]
/e2e (don't worry about that; i'm running a test)

**Comment by auguste-probabl** [comment]
Waiting on #3104 which will help resolve https://github.com/probabl-ai/skore/pull/3099#pullrequestreview-4580086016

File: skore/src/skore/_sklearn/_base.py [hunk 3]
```diff
@@ -179,3 +173,23 @@ def _repr_html_(self) -> str:
 
     def _repr_mimebundle_(self, **kwargs):
         return {"text/plain": repr(self), "text/html": self._repr_html_()}
+
+
+class BaseMetricsAccessor(_BaseAccessor, Generic[ParentT]):
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_base.py:181]: Would you mind moving the `__repr__` and and `_repr_html_` methods that currently live on the 3 different `_MetricsAccessor` here to reduce duplication ?

**Review by GaetandeCast — DISMISSED** [review]
Small request, otherwise LGTM!

File: skore/src/skore/_sklearn/_base.py (hunk 3 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_base.py:181]: done, tell me what you think

File: skore/src/skore/_sklearn/_base.py [hunk 4]
```diff
@@ -172,3 +176,50 @@ def _repr_html_(self) -> str:
...
+
+    def _formatted_summary_frame(
+        self,
+        *,
+        data_source: DataSource = "test",
+        metric: str | list[str] | None = None,
+    ) -> pd.DataFrame:
+        """Metric summary.
+
+        Used for displaying the report.
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_base.py]:
```suggestion
        Used for displaying the report and the accessor. Overwritten in comparion
        report.
```

File: skore/src/skore/_sklearn/_comparison/metrics_accessor.py [hunk 5]
```diff
@@ -119,6 +119,10 @@ def _formatted_summary_frame(
         data_source: DataSource = "test",
         metric: str | list[str] | None = None,
     ) -> pd.DataFrame:
+        """Metric summary.
+
+        Used for displaying the report.
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_comparison/metrics_accessor.py:124]:
```suggestion
        Used for displaying the report and the accessor.
```

File: skore/src/skore/_sklearn/_base.py (hunk 4 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_base.py]: fixed but i don't want to give too much detail that we could forget to update

**Review by GaetandeCast — APPROVED** [review]
Thanks @auguste-probabl !
